File: ios/MenuMakerUITests/PageObjects/CheckoutPage.swift
//
//  CheckoutPage.swift
//  MenuMakerUITests
//
//  Page Object for Checkout Screen
//

import XCTest

struct CheckoutPage {
    let app: XCUIApplication

    // MARK: - Elements

    var deliveryAddressField: XCUIElement {
        app.textFields.matching(NSPredicate(format: "placeholderValue CONTAINS[c] 'address'")).firstMatch
    }

    var paymentMethodButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'card' OR label CONTAINS[c] 'cash' OR label CONTAINS[c] 'upi'"))
    }

    var cardPaymentButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'card'")).firstMatch
    }

    var cashPaymentButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'cash'")).firstMatch
    }

    var upiPaymentButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'upi'")).firstMatch
    }

    var placeOrderButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'place order' OR label CONTAINS[c] 'confirm'")).firstMatch
    }

    var orderSummary: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'order summary'")).firstMatch
    }

    var orderConfirmation: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'order placed' OR label CONTAINS[c] 'confirmed' OR label CONTAINS[c] 'order #'")).firstMatch
    }

    // MARK: - Actions

    @discardableResult
    func enterDeliveryAddress(_ address: String) -> CheckoutPage {
        deliveryAddressField.tap()
        deliveryAddressField.typeText(address)
        return self
    }

    @discardableResult
    func selectPaymentMethod(_ method: PaymentMethod) -> CheckoutPage {
        switch method {
        case .card:
            cardPaymentButton.tap()
        case .cash:
            cashPaymentButton.tap()
        case .upi:
            upiPaymentButton.tap()
        }
        return self
    }

    @discardableResult
    func placeOrder() -> CheckoutPage {
        placeOrderButton.tap()
        return self
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> CheckoutPage {
        XCTAssertTrue(orderSummary.waitForExistence(timeout: timeout) ||
                     placeOrderButton.waitForExistence(timeout: timeout),
                     "Checkout screen should be displayed")
        return self
    }

    @discardableResult
    func assertOrderConfirmed(timeout: TimeInterval = 5) -> CheckoutPage {
        XCTAssertTrue(orderConfirmation.waitForExistence(timeout: timeout), "Order confirmation should be displayed")
        return self
    }

    @discardableResult
    func assertPlaceOrderButtonEnabled() -> CheckoutPage {
        XCTAssertTrue(placeOrderButton.isEnabled, "Place order button should be enabled")
        return self
    }

    // MARK: - Types

    enum PaymentMethod {
        case card
        case cash
        case upi
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/CheckoutPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Checkout Screen
 */
class CheckoutPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val deliveryAddressField = composeTestRule.onNodeWithText("Address", substring = true, ignoreCase = true)
    private val cashPaymentOption = composeTestRule.onNodeWithText("Cash", substring = true, ignoreCase = true)
    private val cardPaymentOption = composeTestRule.onNodeWithText("Card", substring = true, ignoreCase = true)
    private val upiPaymentOption = composeTestRule.onNodeWithText("UPI", substring = true, ignoreCase = true)
    private val placeOrderButton = composeTestRule.onNodeWithText("Place Order", ignoreCase = true)
    private val orderTotal = composeTestRule.onNode(hasTestTag("OrderTotal"))

    // Actions
    fun enterDeliveryAddress(address: String): CheckoutPage {
        deliveryAddressField.performClick()
        deliveryAddressField.performTextInput(address)
        return this
    }

    fun selectPaymentMethod(method: PaymentMethod): CheckoutPage {
        when (method) {
            PaymentMethod.CASH -> cashPaymentOption.performScrollTo().performClick()
            PaymentMethod.CARD -> cardPaymentOption.performScrollTo().performClick()
            PaymentMethod.UPI -> upiPaymentOption.performScrollTo().performClick()
        }
        Thread.sleep(500)
        return this
    }

    fun placeOrder(): CheckoutPage {
        placeOrderButton.performScrollTo()
        placeOrderButton.performClick()
        Thread.sleep(3000)
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): CheckoutPage {
        placeOrderButton.assertExists()
        return this
    }

    fun assertPaymentMethodsDisplayed(): CheckoutPage {
        cashPaymentOption.assertExists()
        return this
    }

    fun assertOrderTotalDisplayed(): CheckoutPage {
        orderTotal.assertExists()
        return this
    }

    enum class PaymentMethod {
        CASH, CARD, UPI
    }
}
